go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -0,0 +1,499 @@
+		}
+
+		if util.NoHostSubnet(node) {
+			if ncc.enableHybridOverlaySubnetAllocator && houtil.IsHybridOverlayNode(node) {
+				// this is a hybrid overlay node so mark as allocated from the hybrid overlay subnet allocator
+				hostSubnet, err := houtil.ParseHybridOverlayHostSubnet(node)
+				if err != nil {
+					klog.Errorf("Failed to parse hybrid overlay for node %s: %w", node.Name, err)
+				} else if hostSubnet != nil {
+					klog.V(5).Infof("Node %s contains subnets: %v", node.Name, hostSubnet)
+					if err := ncc.hybridOverlaySubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnet); err != nil {
+						klog.Errorf("Failed to mark the subnet %v as allocated in the hybrid subnet allocator for node %s: %w", hostSubnet, node.Name, err)
+					}
+				}
+			}
+		} else {
+			hostSubnets, _ := util.ParseNodeHostSubnetAnnotation(node, ncc.networkName)
+			if len(hostSubnets) > 0 {
+				klog.V(5).Infof("Node %s contains subnets: %v for network : %s", node.Name, hostSubnets, ncc.networkName)
+				if err := ncc.clusterSubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnets...); err != nil {
+					klog.Errorf("Failed to mark the subnet %v as allocated in the cluster subnet allocator for node %s: %w", hostSubnets, node.Name, err)
+				}
+			} else {
+				klog.V(5).Infof("Node %s contains no subnets for network : %s", node.Name, ncc.networkName)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (ncc *networkClusterController) updateNodeAnnotationWithRetry(nodeName string, hostSubnetsMap map[string][]*net.IPNet) error {
+	// Retry if it fails because of potential conflict which is transient. Return error in the
+	// case of other errors (say temporary API server down), and it will be taken care of by the
+	// retry mechanism.
+	resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Informer cache should not be mutated, so get a copy of the object
+		node, err := ncc.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return err
+		}
+
+		cnode := node.DeepCopy()
+		for netName, hostSubnets := range hostSubnetsMap {
+			cnode.Annotations, err = util.UpdateNodeHostSubnetAnnotation(cnode.Annotations, hostSubnets, netName)
+			if err != nil {
+				return fmt.Errorf("failed to update node %q annotation subnet %s",
+					node.Name, util.JoinIPNets(hostSubnets, ","))
+			}
+		}
+		return ncc.kube.UpdateNode(cnode)
+	})
+	if resultErr != nil {
+		return fmt.Errorf("failed to update node %s annotation", nodeName)
+	}
+	return nil
+}
+
+// cleanup cleans up the subnet annotations from the node for the network
+func (ncc *networkClusterController) cleanup() error {
+	// remove hostsubnet annotation for this network
+	klog.Infof("Remove node-subnets annotation for network %s on all nodes", ncc.networkName)
+	existingNodes, err := ncc.watchFactory.GetNodes()
+	if err != nil {
+		return fmt.Errorf("error in retrieving the nodes: %v", err)
+	}
+
+	for _, node := range existingNodes {
+		if util.NoHostSubnet(node) {
```

Comment:
I wanted network cluster controller to be generic. I mean it should not ideally check if the network it is managing is default network or secondary network.

I could change it.

One question -  what happens to secondary networks for windows nodes ?  

Lets say we have a deployment with 3 nodes - N1, N2 and N3 and N3 is a windows node (or no host subnet node).  When a secondary layer 3 network is created what happens to this node ?  
With the present master,  a subnet  (from the seconadry layer3 subnet range) is not allocated for N3 right ?

Reply:
Right, we don't support secondary networks on those nodes. I am not sure we are handling it correctly everywhere even if we should. But also we are not doing anything specific to avoid allocating pods attached to secondary networks on those nodes, the user should take care of it for now.

It's confusing that we don't handle in the cleanup all of what the controller could have done. At the same time it makes sense that we don't if in some situations this method shouldn't be called in the first place.

Ack on being generic. But on many places we have compromised the ability to have a common code base for default & secondary networks with specific checks here & there. You can check how we use the method `IsSecondary()` for this.

Reply:
I resolved this by add a check - if !ncc.IsSecondary() { } at the beginning.

networkClusterController struct now embeds util.NetInfo and util.NetConfInfo.

And with this, secondary network cluster manager now returns networkClusterController{} object for secondary layer 3 networks.  We don't need the wrapper 'SecondaryNetworkClusterController' any more.

Thanks